### PR TITLE
Tests: compile output using config files

### LINKED ISSUE #365: Compile output files using a configuration file for options
Transfer issue https://github.com/SwiftGen/templates/issues/75.

Right now our output files are compiled by selecting options for each file based on filename-matching:
- Swift version (2, 3, 4, ...)
- Platform (macOS, iOS, watchOS, tvOS)
- extra definitions
- imported modules
- public/internal access

As mentioned in https://github.com/SwiftGen/templates/pull/36#issuecomment-323175365, it would be nicer to create a system to handle these choices in a clearer way.

### PR DESCRIPTION
**Pull request by djbe**
Fixes #365.

It's been a while since I did anything beyond very basic Ruby stuff, so don't shoot me 😅 Also, naming things is hard...

~Note: this is based on #598, which should be merged first.~

### CODE REVIEW

**Comment by SwiftGen-Eve** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

Hey 👋 I'm Eve, the friendly bot watching over SwiftGen 🤖

Thanks a lot for your contribution!



---

Seems like everything is in order 👍 You did a good job here! 🤝

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Tests/Fixtures/CompilationEnvironment/IB-iOS-NamespacedLogin.swift [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import UIKit
+
+enum ExtraModule {
+  final class SlideDownSegue: UIStoryboardSegue {}
+  final class ValidatePasswordViewController: UIViewController {}
+}
```
Comment (AliSoftware) [inline Tests/Fixtures/CompilationEnvironment/IB-iOS-NamespacedLogin.swift]: missing newline at EOF

File: rakelib/CompilationConfiguration.rb [hunk 2]
```diff
@@ -0,0 +1,116 @@
+require 'yaml'
+
+COMPILATION_CONFIGURATION_FILE = '.compilation-configuration.yml'.freeze
```
Comment (AliSoftware) [inline rakelib/CompilationConfiguration.rb]: I don't think that's worth hiding the file by making it start with `.`. Would make the test look like magic (especially to outside contributors not knowing the history and internals of that compilation-test system)

File: rakelib/CompilationConfiguration.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    'macOS' => Platform.new('macosx', 'x86_64-apple-macosx10.13'),
+    'tvOS' => Platform.new('appletvos', 'arm64-apple-tvos12.0'),
+    'watchOS' => Platform.new('watchos', 'armv7k-apple-watchos5.0')
+  }.freeze
+
+  SWIFT_VERSIONS = [3, 4, 4.2]
+end
+
+class Module
+  def self.commands_for(file, platform)
```
Comment (AliSoftware) [inline rakelib/CompilationConfiguration.rb]: Ruby conventions would probably prefer naming this `commands_for_file(file, platform)`

File: rakelib/CompilationConfiguration.rb [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  def self.load(folder)
+    config_file = "#{folder}#{COMPILATION_CONFIGURATION_FILE}"
+    if File.file?(config_file)
+      YAML.load_file(config_file)
+    else
+      nil
+    end
+  end
+
+  def commands_for(file)
```
Comment (AliSoftware) [inline rakelib/CompilationConfiguration.rb]: Same, `commands_for_file`

**Review by AliSoftware — APPROVED** [review]

**Pull request merged**